python/ray/dag/dag_node.py

```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
+        """Breadth-first search on the DAG."""
+        visited = set()
+        queue = [self]
+        while queue:
+            node = queue.pop(0)
+            if node not in visited:
+                fn(node)
+                visited.add(node)
+                # Add all unseen downstream and upstream nodes to the queue.
+                # This function should be called by the root of the DAG. However,
+                # in some invalid cases, some nodes may not be descendants of the
+                # root. Therefore, we also add upstream nodes to the queue so that
+                # a meaningful error message can be raised when the DAG is compiled.
+                for neighbor in chain.from_iterable(
+                    [node._downstream_nodes, node._upstream_nodes]
+                ):
+                    if neighbor not in visited:
+                        queue.append(neighbor)
+
+    def _find_root(self) -> "DAGNode":
+        """
+        Return the root node of the DAG. The root node must be an InputNode.
+        """
+        from ray.dag.input_node import InputNode
+
+        root = self
+        while not isinstance(root, InputNode):
+            if len(root._upstream_nodes) == 0:
+                raise ValueError("No InputNode found in the DAG.")
+            root = root._upstream_nodes[0]
```

Comment:
assert `len(root._upstream_nodes) == 1`

Reply:
why `len(root._upstream_nodes) == 1`?

Reply:
Right. I misunderstood. I think this just selects an arbitrary upstream node with index 0.